---
title: "Linear models to test significance of ISD between genes controlling for evolutionary rate"
author: "Ben Wilson"
date: "August 9, 2016"
output: html_document
---

Load mySQL table info for genes
```{r}
ISD.data <- read.table("ISD_data.csv",header=T,sep=",") #read mySQL data into data frame
```

Factorize gene family number and phylostratum
```{r}
ISD.data$GeneFamilyNumber <- as.factor(ISD.data$GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
ISD.data$GeneFamilyPhylostratum <- as.factor(ISD.data$GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
ISD.data$GeneFamilyPhylostratum <- with(ISD.data, factor(GeneFamilyPhylostratum, levels = c(sort(unique(ISD.data$GeneFamilyPhylostratum)),"Intergenic","Masked"))) # set levels of categorical phylostratum variable
ISD.data$UnknownFamilyPhylostrataBinary <- as.factor(ISD.data$UnknownFamilyPhylostrataBinary) #factorize binary category for gene families with ambiguous age
ISD.data$ProteinSeqLength <- as.numeric(ISD.data$ProteinSeqLength)
```

Trim data frames to only the information needed
```{r}
ISD.data <- ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene" | ISD.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl" | ISD.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),] #Subset to categories needed for this figure
ISD.data <- ISD.data[,c("EnsemblGeneUID","Phylostratum","GeneFamilyPhylostratum","GeneFamilyNumber","NoCysIUPredMeanISD","GeneOrControlDesignation","EvolutionaryRateHighestOrthologRate","UnknownFamilyPhylostrataBinary", "ProteinSeqLength")] #subset needed columns
ISD.data[which(ISD.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Intergenic" #Give intergenic sequence a categorical phylostratum for comparison
ISD.data[which(ISD.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Masked" #Give repeat-masked intergenic sequence a categorical phylostratum

```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
ISD.data <- ISD.data[which(ISD.data$GeneFamilyPhylostratum != "0" &ISD.data$GeneFamilyPhylostratum != "20" & ISD.data$UnknownFamilyPhylostrataBinary != "1" & !is.na(ISD.data$EvolutionaryRateHighestOrthologRate)),] #exclude unclassifiable genes, false-positive rich ORFans
```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit <- boxcoxfit(ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]$NoCysIUPredMeanISD, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1 <- unname(bcFit$lambda[1]) #extract transformation variables
lambda2 <- unname(bcFit$lambda[2]) #...
lambda1 #print transformation variables
lambda2 #...

bc.transform <- function(x,l1,l2){ #function to transform data according to Box-Cox transformation fit
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

ISD.data$bcISD <- bc.transform(ISD.data$NoCysIUPredMeanISD, lambda1, lambda2) #create another column variable for transformed IUPred ISD values
```

Join old phylostrata where homology detection bias is an issue (pre-vertebrates)
```{r}
ISD.data[which(ISD.data$GeneFamilyPhylostratum %in% 1:10),]$GeneFamilyPhylostratum <- "1"
```

Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}
library(nlme)
fit.ISD.genes <- lme(bcISD ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene"),]) #generate a linear mixed-effects model fit to mouse genes with gene family as random error
summary(fit.ISD.genes) #print fit AIC for mouse genes with gene family random error

fit.ISD.genes.rate.corrected <- lme(bcISD ~ GeneFamilyPhylostratum + EvolutionaryRateHighestOrthologRate, random = ~ 1|GeneFamilyNumber, data = ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]) #generate a linear mixed-effects model fit to mouse genes with gene family as random error, add dNdS as fixed term, exclude genes without a corresponding dNdS value for mouse-rat comparisons
summary(fit.ISD.genes.rate.corrected) #print fit AIC for evolutionary rate corrected linear model

fit.ISD.genes.length.corrected <- lme(bcISD ~ GeneFamilyPhylostratum  + ProteinSeqLength, random = ~ 1|GeneFamilyNumber, data = ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",])#generate fit with length as additional confounder
summary(fit.ISD.genes.length.corrected)


fit.ISD.intergenic <- lm(bcISD ~ 1 , data = ISD.data[ISD.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl",]) #generate a standard linear model for matched intergenic controls

fit.ISD.masked <- lm(bcISD ~ 1 , data = ISD.data[ISD.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl",]) #generate a standard linear model for matched repeat-maksed intergenic controls

```

Import protein length data from length/dNdS analysis
```{r}
protein.lengths <- readRDS("protein_lengths.rds")
length.lambda1 <- readRDS("length_lambda1.rds")
length.lambda2 <- readRDS("length_lambda2.rds")

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

protein.lengths$Lengths <- bc.reverse.transform(protein.lengths$Intercept,length.lambda1,length.lambda2)
vertebrate.lengths <- mean(protein.lengths$Lengths[2:length(protein.lengths$Lengths)])
```

Data frames for plots
```{r}
last.index.genes <- length(fixed.effects(fit.ISD.genes))  #get number of phylostrata that have been fit
genes.df <- data.frame(Phylostratum = names(fixed.effects(fit.ISD.genes)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.ISD.genes)[1:last.index.genes]), SE = unname(summary(fit.ISD.genes)$tTable[,2][1:last.index.genes]), Model = "Genes (uncorrected)") #build data frame to hold linear model data containing PS, mean bcISD, and SEM

last.index.corrected <- length(fixed.effects(fit.ISD.genes.length.corrected)) - 1 #get number of phylostrata that have been fit, minus 1 for fixed effect length
genes.length.corrected.df <- data.frame(Phylostratum = names(fixef(fit.ISD.genes.length.corrected)[1:last.index.corrected]), Intercept = unname(fixef(fit.ISD.genes.length.corrected)[1:last.index.corrected]) + unname(fixef(fit.ISD.genes.length.corrected)[last.index.corrected + 1]) * vertebrate.lengths, SE = unname(summary(fit.ISD.genes.length.corrected)$tTable[,2][1:last.index.corrected]), Model = "Genes length corrected") #build data frame to hold linear model data containing PS, mean bcISD, and SEM


remove(fit.ISD.genes, fit.ISD.genes.length.corrected) #remove linear model objects for genes

intergenic.df <- data.frame(Phylostratum = c("Intergenic"), Intercept = summary(fit.ISD.intergenic)$coefficients[1], SE = summary(fit.ISD.intergenic)$coefficients[2], Model = "Intergenic") #build data frome for intergenic mean bcISD, fix phylostratum (for plotting purposes), add SEM

masked.df <- data.frame(Phylostratum = c("Masked"), Intercept = summary(fit.ISD.masked)$coefficients[1], SE = summary(fit.ISD.masked)$coefficients[2], Model = "Repeat-masked") #...same for repeat-masked intergenic controls

remove(fit.ISD.intergenic, fit.ISD.masked) #clear linear model objects from memory
```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), combine data frames per categorization
```{r}
# Adjust intecepts, combine data frames
genes.df$Intercept[2:last.index.genes] <- genes.df$Intercept[2:last.index.genes] + genes.df$Intercept[1] #increment means by primary intercept, corrects for R convention of relative means

genes.length.corrected.df$Intercept[2:last.index.corrected] <- genes.length.corrected.df$Intercept[2:last.index.corrected] + genes.length.corrected.df$Intercept[1]  #...same as above

combined.df <- data.frame(rbind(genes.df, genes.length.corrected.df, intergenic.df, masked.df)) #combine data frames for plotting
combined.df$Phylostratum <- replace(as.character(combined.df$Phylostratum),1:(last.index.genes+last.index.corrected),c(1:last.index.genes,1:last.index.corrected)) #replace linear model text with readable phylostratum names
combined.df$Phylostratum <- with(combined.df, factor(Phylostratum, levels = c(1:max(last.index.genes,last.index.corrected),"Intergenic","Masked"))) #factorize phylostratum before plotting

```

Create colorblind friendly color scheme 
```{r}
library("ggplot2")
library("grid")

#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]


```

Generate plots for ISD vs Phylostratum
```{r}

plot.ISD.homology <- ggplot(combined.df, aes(colour = Model, fill = Model, shape = Model, alpha = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.ISD.homology <- plot.ISD.homology + 
  geom_linerange(aes(x = Phylostratum, ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), lwd = 0.5, position = position_dodge(width = 0.75), show.legend=FALSE) + #add linerange for SEM bars
  geom_pointrange(aes(x = Phylostratum, y = bc.reverse.transform(Intercept, lambda1, lambda2), ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), size = 0.5, lwd = 0.5, position = position_dodge(width = 0.75),  show.legend =TRUE) + #add points for mean ISD
  #scale_x_discrete(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","Eumetazoa","Bilateria","Deuterostomia","Chordata","Olfactores","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia","","")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_x_discrete(labels=c("Pre-vertebrates","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia","","")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_alpha_manual(name = "", values = c(1.0,0.5,1.0,1.0), guide = guide_legend(ncol=2,byrow=TRUE)) + #make rate-corrected means half transparent
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("ISD") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=9, hjust = 1, vjust = 0, colour = "black"),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.33,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.01,0.01,0.01,0.01), "in"))

plot.ISD.homology

#Create pdf with particular dimensions for journal
pdf(file = "ISDvsPS1-19GenesIntergenicMasked_LengthControl_VertebrateMean.pdf", width =4.6, height = 4)
plot.ISD.homology
dev.off()
```