---
title: "Linear models to test significance of ISD difference between genes and scrambled controls"
author: "Ben Wilson"
date: "August 9, 2016"
output: html_document
---
  
Load mySQL table info for genes
```{r}
ISD.data <- read.table("ISD_data.csv",header=T,sep=",") #read mySQL data into data frame
```

Factorize gene family number and phylostratum
```{r}
ISD.data$GeneFamilyNumber <- as.factor(ISD.data$GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
ISD.data$GeneFamilyPhylostratum <- as.factor(ISD.data$GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
ISD.data$GeneFamilyPhylostratum <- with(ISD.data, factor(GeneFamilyPhylostratum, levels = c(sort(unique(ISD.data$GeneFamilyPhylostratum))))) # set levels of categorical phylostratum variable
ISD.data$UnknownFamilyPhylostrataBinary <- as.factor(ISD.data$UnknownFamilyPhylostrataBinary) #factorize binary category for gene families with ambiguous age
```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, and ORFan proteins
```{r}
ISD.data <- ISD.data[which(ISD.data$GeneFamilyPhylostratum != "0" & ISD.data$GeneFamilyPhylostratum != "20" & ISD.data$UnknownFamilyPhylostrataBinary != "1" & !is.na(ISD.data$EvolutionaryRateHighestOrthologRate)),] #exclude unclassifiable genes, false-positive rich ORFans
```

Join old phylostrata where homology detection bias is an issue (pre-vertebrates)
```{r}
ISD.data[which(ISD.data$GeneFamilyPhylostratum %in% 1:10),]$GeneFamilyPhylostratum <- "1"
```

Trim data frames to only the information needed, split into categories
```{r}
ISD.data <- ISD.data[,c("EnsemblGeneUID","EnsemblProteinUID","GeneFamilyPhylostratum","GeneFamilyNumber","NoCysIUPredMeanISD","GeneOrControlDesignation","UnknownFamilyPhylostrataBinary")] #subset needed columns

genes <- ISD.data[which(ISD.data$GeneOrControlDesignation=="CodingGene"),c("EnsemblGeneUID","GeneFamilyPhylostratum","GeneFamilyNumber","NoCysIUPredMeanISD")] #split out genes
scrambled <- ISD.data[which(ISD.data$GeneOrControlDesignation=="ScrambledByAminoAcidControl"), c("EnsemblGeneUID","NoCysIUPredMeanISD")] #...scrambled controls,
gc <- ISD.data[which(ISD.data$GeneOrControlDesignation=="RandomlyGeneratedControlWithConsistentGC"), c("EnsemblGeneUID","NoCysIUPredMeanISD")] #...and GC controls

remove(ISD.data) #clear bulk data frame

colnames(genes)[4] <- "GeneNoCysMeanISD" #make ISD column names unique so that they can be merged by EnsemblGeneUID
colnames(scrambled)[2] <- "ScrambledNoCysMeanISD" #same here
colnames(gc)[2] <- "GCNoCysMeanISD" #same here

```

Merge genes and paired controls by EnsemblGeneID, combine into single data frame for calculating pairwise difference
```{r}
paired.ISD.df <- merge(genes, scrambled, by = "EnsemblGeneUID", all.x = TRUE) #merge genes and scrambled first
ISD.df <- merge(paired.ISD.df, gc, by = "EnsemblGeneUID", all.x = TRUE) #merge the previous with GC controls (R merge only allows pairwise mergers)
remove(genes) 
remove(scrambled) 
remove(gc)
remove(paired.ISD.df)
```

Construct comparison for the difference between means of scrambled and unscrambled genes and GC controls
```{r}
#Construct a data frame for the difference in means between genes and scrambled genes
ISD.df$DifferenceGeneVsScrambledISD <- ISD.df$GeneNoCysMeanISD - ISD.df$ScrambledNoCysMeanISD
ISD.df$DifferenceGeneVsGCISD <- ISD.df$GeneNoCysMeanISD - ISD.df$GCNoCysMeanISD
```

Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}
library(nlme)
lm.ISD.genes.scrambled <- lme(DifferenceGeneVsScrambledISD ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = ISD.df) #linear mixed-effects model for scrambled difference with gene families as independent sources of error
lm.ISD.genes.gc <- lme(DifferenceGeneVsGCISD ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data =ISD.df) #linear mixed-effects model for GC difference with gene families as independent sources of error

remove(ISD.df)
```

Plot means from each model with confidence intervals to test if they are significantly different
```{r}
# Put model estimates into temporary data frames
last.index.scrambled <- length(fixed.effects(lm.ISD.genes.scrambled)) #get number of phylostrata that have been fit
genes.scrambled.df <- data.frame(Phylostratum = names(fixed.effects(lm.ISD.genes.scrambled)[1:last.index.scrambled]), Intercept = unname(fixed.effects(lm.ISD.genes.scrambled)[1:last.index.scrambled]), SE = unname(summary(lm.ISD.genes.scrambled)$tTable[,2][1:last.index.scrambled]), Model = "Genes minus Scrambled") #construct data frame to hold linear model results
genes.scrambled.df$Intercept[2:last.index.scrambled] <- genes.scrambled.df$Intercept[2:last.index.scrambled] + genes.scrambled.df$Intercept[1] #increment means by first intercept to correct for R convention of relative means

last.index.gc <- length(fixed.effects(lm.ISD.genes.gc)) #same for GC controls...
genes.GC.df <- data.frame(Phylostratum = names(fixef(lm.ISD.genes.gc)[1:last.index.gc]), Intercept = unname(fixef(lm.ISD.genes.gc)[1:last.index.gc]), SE = unname(summary(lm.ISD.genes.gc)$tTable[,2][1:last.index.gc]), Model = "Genes minus GC") #...
genes.GC.df$Intercept[2:last.index.gc] <- genes.GC.df$Intercept[2:last.index.gc] + genes.GC.df$Intercept[1] #...

genes.scrambled.df$Phylostratum <- replace(as.character(genes.scrambled.df$Phylostratum),1:last.index.scrambled,c(1:last.index.scrambled)) #replace linear model text with factorized phylostrata
genes.scrambled.df$Phylostratum <- with(genes.scrambled.df, factor(Phylostratum, levels = c(1:last.index.scrambled)))

genes.GC.df$Phylostratum <- replace(as.character(genes.GC.df$Phylostratum),1:last.index.gc,c(1:last.index.gc)) #same for GC data frame
genes.GC.df$Phylostratum <- with(genes.GC.df, factor(Phylostratum, levels = c(1:last.index.gc)))

```

Create colorblind friendly color scheme and reverse transform function
```{r}

#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

```

Make plots to visualize differences in ISD
```{r}
# Plots for each categorization
library("ggplot2")
library("grid")
plot.ISD.diff.scrambled <- ggplot(genes.scrambled.df, aes(colour = Model, fill=Model))
plot.ISD.diff.scrambled <- plot.ISD.diff.scrambled + 
  geom_hline(yintercept = 0, linetype = "dashed", colour = "black") +
  geom_linerange(aes(x = Phylostratum, ymin = Intercept - 1.96*SE, ymax = Intercept + 1.96*SE), lwd = 0.5, show.legend = FALSE) +
  geom_pointrange(aes(x = Phylostratum, y = Intercept, ymin = Intercept - 1.96*SE, ymax = Intercept + 1.96*SE), size = 0.5, lwd = 0.5,  shape = 21, show.legend = TRUE) +
  #scale_x_discrete(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","Eumetazoa","Bilateria","Deuterostomia","Chordata","Olfactores","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia")) +
  scale_x_discrete(labels=c("Pre-vertebrates","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia","","")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_shape_manual(name = "", values = c(22)) +
  scale_colour_manual(name = "", values = c(colVermillion)) +
  scale_fill_manual(name = "", values = c(colVermillion)) +
  #ylim(-0.01,0.025) +
  ylab("Paired difference in ISD") +
  theme(#plot.title = element_text(size=12),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11, colour="black"),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=9, hjust = 1, vjust = 0, colour = "black"),
        legend.title = element_text(size=9),
        legend.text = element_text(size=9),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.41,0.85),
        legend.direction = "horizontal",
        #panel.grid.major = element_blank(),
        plot.margin=unit(c(0.01,0.01,0.01,0.01), "in"))


lambda1 <- readRDS("lambda1.rds")
lambda2 <- readRDS("lambda2.rds")
p <- readRDS("ISDvsPhylostratumPlotObject.rds")
grobp <- ggplotGrob(p + annotate("text",x = 1, y = 0.65,label="A",size=5,fontface="bold")) 

grobp1 <- ggplotGrob(plot.ISD.diff.scrambled + annotate("text",x=1,y=0.045,label="B",size=5,fontface="bold"))
#grobp2<- ggplotGrob(p2 + annotate("text",x=1,y=0.24,label="C",size=5,fontface="bold" ))

pdf(file = "PairedDifferenceISDvsPS1-19_OnlyRatHomologs.pdf", width= 4.6, height = 8)
grid.draw(rbind(grobp,grobp1,size="last"))
dev.off()
grid.draw(rbind(grobp,grobp1,size="last"))
```
