---
title: "Linear models to test significance of ISD between genes and various controls"
author: "Ben Wilson"
date: "August 9, 2016"
output: html_document
---
  
Load mySQL table info for genes
```{r}
ISD.data <- read.table("ISD_data.csv",header=T,sep=",") #read mySQL data into data frame
```

Factorize gene family number and phylostratum
```{r}
ISD.data$GeneFamilyNumber <- as.factor(ISD.data$GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
ISD.data$GeneFamilyPhylostratum <- as.factor(ISD.data$GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
ISD.data$GeneFamilyPhylostratum <- with(ISD.data, factor(GeneFamilyPhylostratum, levels = c(sort(unique(ISD.data$GeneFamilyPhylostratum)),"Intergenic","Masked"))) # set levels of categorical phylostratum variable
ISD.data$UnknownFamilyPhylostrataBinary <- as.factor(ISD.data$UnknownFamilyPhylostrataBinary) #factorize binary category for gene families with ambiguous age
```

Trim data frames to only the information needed
```{r}
ISD.data <- ISD.data[,c("EnsemblGeneUID","EnsemblProteinUID","Phylostratum","GeneFamilyPhylostratum","GeneFamilyNumber","NoCysIUPredMeanISD","GeneOrControlDesignation","UnknownFamilyPhylostrataBinary","EvolutionaryRateHighestOrthologRate")] #subset needed columns
ISD.data[which(ISD.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Intergenic" #Give intergenic sequence a categorical phylostratum for comparison
ISD.data[which(ISD.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl"),]$GeneFamilyPhylostratum <- "Masked" #Give repeat-masked intergenic sequence a categorical phylostratum
```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
ISD.data <- ISD.data[which(ISD.data$GeneFamilyPhylostratum != "0" & ISD.data$GeneFamilyPhylostratum != "20" & ISD.data$UnknownFamilyPhylostrataBinary != "1" & !is.na(ISD.data$EvolutionaryRateHighestOrthologRate)),] #exclude unclassifiable genes, false-positive rich ORFans
```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit <- boxcoxfit(ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]$NoCysIUPredMeanISD, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1 <- unname(bcFit$lambda[1]) #extract transformation variables
lambda2 <- unname(bcFit$lambda[2]) #...
lambda1 #print transformation variables
lambda2 #...

bc.transform <- function(x,l1,l2){ #function to transform data according to Box-Cox transformation fit
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

ISD.data$bcISD <- bc.transform(ISD.data$NoCysIUPredMeanISD, lambda1, lambda2) #create another column variable for transformed IUPred ISD values
```

Join old phylostrata where homology detection bias is an issue (pre-vertebrates)
```{r}
ISD.data[which(ISD.data$GeneFamilyPhylostratum %in% 1:10),]$GeneFamilyPhylostratum <- "1"
```

Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}
library(nlme)
fit.ISD.genes.standard <- lm(bcISD ~ GeneFamilyPhylostratum, data = ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene"),]) #fit standard linear model of ISD vs. PS
genes.independent.AIC <- AIC(fit.ISD.genes.standard) #print AIC
summary(fit.ISD.genes.standard)

fit.ISD.genes <- lme(bcISD ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene"),]) #fit mixed-effects model with gene family annotation as random effect
full.model.AIC <- summary(fit.ISD.genes)$AIC #print AIC
summary(fit.ISD.genes)

fit.ISD.genes.random <- lme(bcISD ~ 1, random = ~ 1|GeneFamilyNumber, data = ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene"),]) #fit model with PS and just random effect of gene family for comparison
no.phylostratum.AIC <- summary(fit.ISD.genes.random)$AIC #print AIC
summary(fit.ISD.genes.random)

remove(fit.ISD.genes.standard)
remove(fit.ISD.genes.random)

print("Delta AIC genes independent versus gene families independent: ")
genes.independent.AIC - full.model.AIC
print("Delta AIC no phylostratum versus gene families independent: ")
no.phylostratum.AIC - full.model.AIC

fit.ISD.scrambled <- lme(bcISD ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = ISD.data[ISD.data$GeneOrControlDesignation == "ScrambledByAminoAcidControl",]) #fit mixed-effects model to scrambled controls with random error on gene families

fit.ISD.gc <- lme(bcISD ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = ISD.data[ISD.data$GeneOrControlDesignation == "RandomlyGeneratedControlWithConsistentGC" ,]) #fit mixed-effects model to GC-content controls with random error on gene families

fit.ISD.intergenic <- lm(bcISD ~ 1, data = ISD.data[ISD.data$GeneOrControlDesignation == "IntergenicNucleotideSeqControl",]) #fit standard linear model to intergenic controls

fit.ISD.masked <- lm(bcISD ~ 1, data = ISD.data[ISD.data$GeneOrControlDesignation == "MaskedIntergenicNucleotideSeqControl",]) #fit standard linear model to repeat-masked intergenic controls

remove(ISD.data) #clear object containing modified raw data
```

Plot means from each model with confidence intervals to test if they are significantly different
```{r}
# Put model estimates into data frames
last.index.genes <- length(fixed.effects(fit.ISD.genes)) #get number of phylostrata that have been fit
genes.df <- data.frame(Phylostratum = names(fixed.effects(fit.ISD.genes)[1:last.index.genes]), Intercept = unname(fixed.effects(fit.ISD.genes)[1:last.index.genes]), SE = unname(summary(fit.ISD.genes)$tTable[,2][1:last.index.genes]), Model = "Genes") #build data frame to hold linear model data containing PS, mean bcISD, and SEM
remove(fit.ISD.genes)


last.index.scrambled <- length(fixed.effects(fit.ISD.scrambled))#get number of phylostrata that have been fit
scrambled.df <- data.frame(Phylostratum = names(fixed.effects(fit.ISD.scrambled)[1:last.index.scrambled]), Intercept = unname(fixed.effects(fit.ISD.scrambled)[1:last.index.scrambled]), SE = unname(summary(fit.ISD.scrambled)$tTable[,2][1:last.index.scrambled]), Model = "Scrambled") #build data frame to hold linear model data containing PS, mean bcISD, and SEM
remove(fit.ISD.scrambled)


last.index.gc <- length(fixed.effects(fit.ISD.gc)) #get number of phylostrata that have been fit
gc.df <- data.frame(Phylostratum = names(fixed.effects(fit.ISD.gc)[1:last.index.gc]), Intercept = unname(fixed.effects(fit.ISD.gc)[1:last.index.gc]), SE = unname(summary(fit.ISD.gc)$tTable[,2][1:last.index.gc]), Model = "GC") #build data frame to hold linear model data containing PS, mean bcISD, and SEM
remove(fit.ISD.gc)


intergenic.df <- data.frame(Phylostratum = c("Intergenic"), Intercept = summary(fit.ISD.intergenic)$coefficients[1], SE = summary(fit.ISD.intergenic)$coefficients[2], Model = "Intergenic") #build data frome for intergenic mean bcISD, fix phylostratum (for plotting purposes), add SEM

masked.df <- data.frame(Phylostratum = c("Repeat-masked"), Intercept = summary(fit.ISD.masked)$coefficients[1], SE = summary(fit.ISD.masked)$coefficients[2], Model = "Repeat-masked") #...same for repeat-masked intergenic controls
```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), combine data frames per categorization
```{r}
# Adjust intecepts, combine data frames

genes.df$Intercept[2:last.index.genes] <- genes.df$Intercept[2:last.index.genes] + genes.df$Intercept[1] #increment means by primary intercept, corrects for R convention of means relative to the first


scrambled.df$Intercept[2:last.index.scrambled] <- scrambled.df$Intercept[2:last.index.scrambled] + scrambled.df$Intercept[1] #...same for scrambled


gc.df$Intercept[2:last.index.gc] <- gc.df$Intercept[2:last.index.gc] + gc.df$Intercept[1] #...same for GC controls


combined.df <- data.frame(rbind(genes.df, scrambled.df, gc.df, intergenic.df, masked.df)) #combine data frames for plotting
combined.df$Phylostratum <- replace(as.character(combined.df$Phylostratum),1:(last.index.genes+last.index.scrambled+last.index.gc),c(1:last.index.genes,1:last.index.scrambled,1:last.index.gc))
combined.df$Phylostratum <- with(combined.df, factor(Phylostratum, levels = c(1:max(last.index.genes,last.index.scrambled,last.index.gc),"Intergenic","Masked")))

```

Create colorblind friendly color scheme and reverse transform function
```{r}

#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}

```

Make plots to visualize effects of different controls versus gene data
```{r}
# Plots for each categorization
library("ggplot2")
library("grid")

#Create ggplot2 plotting object to create plot with various controls
plot.ISD <- ggplot(combined.df, aes(colour = Model, fill = Model, shape = Model))
plot.ISD <- plot.ISD + 
  geom_linerange(aes(x = Phylostratum, ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), lwd = 0.5, position = position_dodge(width = 1), show.legend=FALSE) +
  geom_pointrange(aes(x = Phylostratum, y = bc.reverse.transform(Intercept, lambda1, lambda2), ymin = bc.reverse.transform(Intercept - SE, lambda1, lambda2), ymax = bc.reverse.transform(Intercept + SE, lambda1, lambda2)), size = 0.5, lwd = 0.5, position = position_dodge(width = 1),  show.legend=TRUE) +
  #scale_x_discrete(labels=c("Cellular org.","Eukaryota","Opisthokonta","Holozoa","Metazoa","Eumetazoa","Bilateria","Deuterostomia","Chordata","Olfactores","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia","","")) +
  scale_x_discrete(labels=c("Pre-vertebrates","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia","","")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_shape_manual(name = "", values = c(21,22,23,24,25)) +
  scale_colour_manual(name = "", values = c(colBlack,colVermillion,colOrange,colDarkBlue,colSkyBlue)) +
  scale_fill_manual(name = "", values = c(colBlack,colVermillion,colOrange,colDarkBlue,colSkyBlue)) +
  ylab("ISD") +
  guides(shape=guide_legend(ncol=3, byrow=TRUE)) +
  theme(axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=9, hjust = 1, vjust = 0, colour = "black"),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.30,0.83),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.01,0.01,0.01,0.01), "in"))

plot.ISD

#Create pdf with particular dimensions 
pdf(file = "ISDvsPS1-19GenesScrambledGCIntergenicMasked_OnlyRatHomologs.pdf", width =4.6, height = 3)
plot.ISD
dev.off()
plot.ISD
saveRDS(lambda1, "lambda1.rds")
saveRDS(lambda2, "lambda2.rds")
saveRDS(plot.ISD, "ISDvsPhylostratumPlotObject.rds")
```