---
title: "Linear models for length and dNdS versus phylostratum"
author: "Ben Wilson"
date: "August 29, 2016"
output: html_document
---

Load mySQL table info for genes
```{r}
ISD.data <- read.table("ISD_data.csv",header=T,sep=",") #read mySQL data into data frame
```

Factorize gene family number and phylostratum
```{r}
ISD.data$GeneFamilyNumber <- as.factor(ISD.data$GeneFamilyNumber) #factorize gene family ID to prevent numeric interpretation by R
ISD.data$GeneFamilyPhylostratum <- as.factor(ISD.data$GeneFamilyPhylostratum) #factorize variable to ensure that it is not interpreted as numeric variable
ISD.data$GeneFamilyPhylostratum <- with(ISD.data, factor(GeneFamilyPhylostratum, levels = c(sort(unique(ISD.data$GeneFamilyPhylostratum)),"Intergenic","Masked"))) # set levels of categorical phylostratum variable
ISD.data$UnknownFamilyPhylostrataBinary <- as.factor(ISD.data$UnknownFamilyPhylostrataBinary) #factorize binary category for gene families with ambiguous age
ISD.data$ProteinSeqLength <- as.numeric(ISD.data$ProteinSeqLength)
```

Trim data frames to only the information needed
```{r}
ISD.data <- ISD.data[,c("EnsemblGeneUID","GeneFamilyPhylostratum","GeneFamilyNumber","NoCysIUPredMeanISD","GeneOrControlDesignation","EvolutionaryRateHighestOrthologRate","UnknownFamilyPhylostrataBinary", "ProteinSeqLength")] #subset needed columns

```

Remove phylostratum 0 (unclassifiable), gene families with ambiguous age, genes without rat homologs, and ORFan proteins
```{r}
ISD.data <- ISD.data[which(ISD.data$GeneFamilyPhylostratum != "0" &ISD.data$GeneFamilyPhylostratum != "20" & ISD.data$UnknownFamilyPhylostrataBinary != "1" & !is.na(ISD.data$EvolutionaryRateHighestOrthologRate)),] #exclude unclassifiable genes, false-positive rich ORFans
```

Use two-parameter Box-Cox transform to reduce skewness
```{r}
library(geoR)
bcFit.length <- boxcoxfit(ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]$ProteinSeqLength, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1.length <- unname(bcFit.length$lambda[1]) #extract transformation variables
lambda2.length <- unname(bcFit.length$lambda[2]) #...
lambda1.length #print transformation variables
lambda2.length #...

bc.transform <- function(x,l1,l2){ #function to transform data according to Box-Cox transformation fit
  x.transform <- ((x+l2)^l1 - 1)/l1
  return(x.transform)
}

ISD.data$bcLength <- bc.transform(ISD.data$ProteinSeqLength, lambda1.length, lambda2.length) #create another column variable for transformed  values

bcFit.rate <- boxcoxfit(ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]$EvolutionaryRateHighestOrthologRate, lambda2 =  TRUE) #fit a Box-Cox transform to data, exclude mouse-specific genes
lambda1.rate <- unname(bcFit.rate$lambda[1]) #extract transformation variables
lambda2.rate <- unname(bcFit.rate$lambda[2]) #...
lambda1.rate #print transformation variables
lambda2.rate #...

ISD.data$bcRate <- bc.transform(ISD.data$EvolutionaryRateHighestOrthologRate, lambda1.rate, lambda2.rate) #create another column variable for transformed values
```

Join old phylostrata where homology detection bias is an issue (pre-vertebrates)
```{r}
ISD.data[which(ISD.data$GeneFamilyPhylostratum %in% 1:10),]$GeneFamilyPhylostratum <- "1"
```

Generate family of linear models for each phylostratum, test significance of gene vs. randomized control controlling for gene family variance
```{r}
library(nlme)
fit.length.genes <- lme(bcLength ~ GeneFamilyPhylostratum , random = ~ 1|GeneFamilyNumber, data = ISD.data[which(ISD.data$GeneOrControlDesignation == "CodingGene"),]) #generate a linear mixed-effects model fit to mouse genes with gene family as random error
summary(fit.length.genes) #print fit AIC for mouse genes with gene family random error

fit.rate.genes <- lme(bcRate ~ GeneFamilyPhylostratum, random = ~ 1|GeneFamilyNumber, data = ISD.data[ISD.data$GeneOrControlDesignation == "CodingGene",]) #generate a linear mixed-effects model fit to mouse genes with gene family as random error, add dNdS as fixed term, exclude genes without a corresponding dNdS value for mouse-rat comparisons
summary(fit.rate.genes) #print fit AIC for evolutionary rate corrected linear model

remove(ISD.data)
```

Data frames for plots
```{r}
last.index.length <- length(fixed.effects(fit.length.genes))  #get number of phylostrata that have been fit
length.df <- data.frame(Phylostratum = names(fixed.effects(fit.length.genes)[1:last.index.length]), Intercept = unname(fixed.effects(fit.length.genes)[1:last.index.length]), SE = unname(summary(fit.length.genes)$tTable[,2][1:last.index.length]), Model = "Length") #build data frame to hold linear model data containing PS, mean bcLength, and SEM

last.index.rate <- length(fixed.effects(fit.rate.genes)) #get number of phylostrata that have been fit
rate.df <- data.frame(Phylostratum = names(fixef(fit.rate.genes)[1:last.index.rate]), Intercept = unname(fixef(fit.rate.genes)[1:last.index.rate]), SE = unname(summary(fit.rate.genes)$tTable[,2][1:last.index.rate]), Model = "Evolutionary rate") #build data frame to hold linear model data containing PS, mean bcISD, and SEM

remove(fit.length.genes, fit.rate.genes) #remove linear model objects for genes
```

Adjust means for each phylostrata by incrementing by the first mean (adjust for R convention), combine data frames per categorization
```{r}
# Adjust intecepts, combine data frames
length.df$Intercept[2:last.index.length] <- length.df$Intercept[2:last.index.length] + length.df$Intercept[1] #increment means by primary intercept, corrects for R convention of relative means
length.df$Phylostratum <- replace(as.character(length.df$Phylostratum),1:(last.index.length),c(1:last.index.length)) #replace linear model text with readable phylostratum names
length.df$Phylostratum <- with(length.df, factor(Phylostratum, levels = c(1:last.index.length)))#spcify factor levels

rate.df$Intercept[2:last.index.rate] <- rate.df$Intercept[2:last.index.rate] + rate.df$Intercept[1]  #...same as above
rate.df$Phylostratum <- replace(as.character(rate.df$Phylostratum),1:(last.index.rate),c(1:last.index.rate)) #replace linear model text with readable phylostratum names
rate.df$Phylostratum <- with(rate.df, factor(Phylostratum, levels = c(1:last.index.rate)))#spcify factor levels
```

Create colorblind friendly color scheme and reverse transform function
```{r}
library("ggplot2")
library("grid")

#Create colorblind friendly palette
colorblindPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colBlack <- colorblindPalette[1]
colOrange <- colorblindPalette[2]
colSkyBlue <- colorblindPalette[3]
colGreen <- colorblindPalette[4]
colYellow <- colorblindPalette[5]
colDarkBlue <- colorblindPalette[6]
colVermillion <- colorblindPalette[7]
colPink <- colorblindPalette[8]

#Create function for reverse box-cox transform to retransform data onto readable axis (note that this preserves relative comparison of certainty between PS but does not preserve absolute measure of SEM)
bc.reverse.transform <- function(x,l1,l2){
  x.reverse.transform <- (x*l1+1)^(1/l1)-l2
  return(x.reverse.transform)
}
```

Generate plots for length/evolutionary-rate vs Phylostratum
```{r}
saveRDS(length.df, "protein_lengths.rds")
saveRDS(lambda1.length, "length_lambda1.rds")
saveRDS(lambda2.length, "length_lambda2.rds")
plot.length <- ggplot(length.df, aes(colour = Model, fill = Model, shape = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.length <- plot.length + 
  geom_linerange(aes(x = Phylostratum, ymin = bc.reverse.transform(Intercept - SE, lambda1.length, lambda2.length), ymax = bc.reverse.transform(Intercept + SE, lambda1.length, lambda2.length)), lwd = 0.5, show.legend=FALSE) + #add linerange for SEM bars
  geom_pointrange(aes(x = Phylostratum, y = bc.reverse.transform(Intercept, lambda1.length, lambda2.length), ymin = bc.reverse.transform(Intercept - SE, lambda1.length, lambda2.length), ymax = bc.reverse.transform(Intercept + SE, lambda1.length, lambda2.length)), size = 0.5, lwd = 0.5,  show.legend =TRUE) + #add points for mean ISD
  scale_x_discrete(labels=c("Pre-vertebrates","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("Protein length (aa)") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=9, hjust = 1, vjust = 0, colour = "black"),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.8,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.01,0.01,0.01,0.01), "in"))

plot.length

#Create pdf with particular dimensions for journal
pdf(file = "LengthVsPhylostratum.pdf", width =4.6, height = 4)
plot.length
dev.off()

plot.rate <- ggplot(rate.df, aes(colour = Model, fill = Model, shape = Model)) #make ggplot plot object for linear model data frame, exclude mouse-specific proteins
plot.rate <- plot.rate + 
  geom_linerange(aes(x = Phylostratum, ymin = bc.reverse.transform(Intercept - SE, lambda1.rate, lambda2.rate), ymax = bc.reverse.transform(Intercept + SE, lambda1.rate, lambda2.rate)), lwd = 0.5, show.legend=FALSE) + #add linerange for SEM bars
  geom_pointrange(aes(x = Phylostratum, y = bc.reverse.transform(Intercept, lambda1.rate, lambda2.rate), ymin = bc.reverse.transform(Intercept - SE, lambda1.rate, lambda2.rate), ymax = bc.reverse.transform(Intercept + SE, lambda1.rate, lambda2.rate)), size = 0.5, lwd = 0.5,  show.legend =TRUE) + #add points for mean ISD
  scale_x_discrete(labels=c("Pre-vertebrates","Vertebrata","Euteleostomi","Tetrapoda","Amniota","Mammalia","Eutheria","Boreoeutheria","Euarchontoglires","Rodentia")) + #change phylostratum labels to correspond to taxonomic nodes
  scale_colour_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue), guide = guide_legend(ncol=2, byrow=TRUE)) + #set colors for each category
  scale_fill_manual(name = "", values = c(colBlack,colGreen,colDarkBlue,colSkyBlue)) + #set fills for each category
  scale_shape_manual(name = "", values = c(21,22,23,24)) + #set shapes for each category
  ylab("Evolutionary rate (dN/dS)") + #set y-axis label
  theme( #set font sizes, orientations, justifications, etc.
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=9, colour="black"),
        axis.title.y = element_text(size=11),
        axis.ticks = element_blank(),
        axis.text.x=element_text(angle = 90, size=9, hjust = 1, vjust = 0, colour = "black"),
        legend.title = element_text(size=9),
        legend.text = element_text(size=8),
        legend.background = element_rect(fill="transparent"),
        legend.position = c(0.33,0.8),
        legend.direction = "horizontal",
        plot.margin=unit(c(0.01,0.01,0.01,0.01), "in"))

plot.rate

#Create pdf with particular dimensions for journal
pdf(file = "EvolutionaryRateVsPhylostratum.pdf", width =4.6, height = 4)
plot.rate
dev.off()


#Create stacked plot to show both correlations
ggGrobLength <- ggplotGrob(plot.length)
ggGrobRate <- ggplotGrob(plot.rate)
grid.draw(rbind(ggGrobLength,ggGrobRate,size="last"))
pdf(file = "LengthPlusRateVsPhylostratum.pdf", width = 4.6, height = 8)
grid.draw(rbind(ggGrobLength,ggGrobRate,size="last"))
dev.off()
```